exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,226 @@
+	r.once.Do(func() {
+		if r.epsListWatcher != nil {
+			r.logger.Debug("creating and starting endpoints informer")
+			epsInformer := cache.NewSharedInformer(r.epsListWatcher, &corev1.Endpoints{}, 0)
+			if _, err := epsInformer.AddEventHandler(r.handler); err != nil {
+				r.logger.Error("unable to start watching for changes to the specified service names", zap.Error(err))
+			}
+			go epsInformer.Run(r.stopCh)
+			if !cache.WaitForCacheSync(r.stopCh, epsInformer.HasSynced) {
+				initErr = errors.New("endpoints informer not sync")
+			}
+		}
+	})
+	if initErr != nil {
+		return initErr
+	}
+
+	r.logger.Debug("K8s service resolver started",
+		zap.String("service", r.svcName),
+		zap.String("namespace", r.svcNs),
+		zap.Int32s("ports", r.port))
+	return nil
+}
+
+func (r *k8sResolver) shutdown(_ context.Context) error {
+	r.changeCallbackLock.Lock()
+	r.onChangeCallbacks = nil
+	r.changeCallbackLock.Unlock()
+
+	close(r.stopCh)
+	r.shutdownWg.Wait()
+	return nil
+}
+func newInClusterClient() (kubernetes.Interface, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(cfg)
+}
+
+func (r *k8sResolver) resolve(ctx context.Context) ([]string, error) {
+	r.shutdownWg.Add(1)
+	defer r.shutdownWg.Done()
+
+	var backends []string
+	r.endpointsStore.Range(func(address, value any) bool {
+		addr := address.(string)
+		if len(r.port) == 0 {
+			backends = append(backends, addr)
+		} else {
+			for _, port := range r.port {
+				backends = append(backends, net.JoinHostPort(addr, strconv.FormatInt(int64(port), 10)))
+			}
+		}
+		return true
+	})
+	_ = stats.RecordWithTags(ctx, k8sResolverSuccessTrueMutators, mNumResolutions.M(1))
+
+	// keep it always in the same order
+	sort.Strings(backends)
+
+	if slices.Equal(r.Endpoints(), backends) {
+		return r.Endpoints(), nil
+	}
+
+	// the list has changed!
+	r.updateLock.Lock()
+	r.endpoints = backends
+	r.updateLock.Unlock()
+	_ = stats.RecordWithTags(ctx, k8sResolverSuccessTrueMutators, mNumBackends.M(int64(len(backends))))
+
+	// propagate the change
+	r.changeCallbackLock.RLock()
+	for _, callback := range r.onChangeCallbacks {
+		callback(r.Endpoints())
+	}
+	r.changeCallbackLock.RUnlock()
+	return r.Endpoints(), nil
+}
+
+func (r *k8sResolver) onChange(f func([]string)) {
+	r.changeCallbackLock.Lock()
+	defer r.changeCallbackLock.Unlock()
+	r.onChangeCallbacks = append(r.onChangeCallbacks, f)
+}
+func (r *k8sResolver) Endpoints() []string {
+	r.updateLock.RLock()
+	defer r.updateLock.RUnlock()
+	return r.endpoints
+}
+
+const inClusterNamespacePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+
+func getInClusterNamespace() (string, error) {
+	// Check whether the namespace file exists.
+	// If not, we are not running in cluster so can't guess the namespace.
+	if _, err := os.Stat(inClusterNamespacePath); os.IsNotExist(err) {
+		return "", fmt.Errorf("not running in-cluster, please specify LeaderElectionNamespace")
```

Comment:
LeaderElectionNamespace?

Reply:
ok,  maybe we just need to use namespaces and "LeaderElectionNamespace" has no role in this context.

Reply:
maybe @jpkrohling thought it was a typo here.

Reply:
done